src/jit/importer.cpp

```diff
@@ -18393,4 +18399,366 @@ bool Compiler::IsMathIntrinsic(GenTreePtr tree)
+    if ((baseClassAttribs & CORINFO_FLG_INTERFACE) != 0)
+    {
+        assert(call->IsVirtualStub());
+        JITDUMP("--- base class is interface, sorry\n");
+        return;
+    }
+
+    // Bail if obj class is an interface.
+    // See for instance System.ValueTuple`8::GetHashCode, where lcl 0 is System.IValueTupleInternal
+    //   IL_021d:  ldloc.0
+    //   IL_021e:  callvirt   instance int32 System.Object::GetHashCode()
+    if ((objClassAttribs & CORINFO_FLG_INTERFACE) != 0)
+    {
+        JITDUMP("--- obj class is interface, sorry\n");
+        return;
+    }
+
+    // Fetch the method that would be called based on the declared type of 'this'
+    CORINFO_METHOD_HANDLE derivedMethod = info.compCompHnd->resolveVirtualMethod(baseMethod, objClass);
+
+    // If we failed to get a handle, we can't devirtualize.  This can
+    // happen when prejitting, if the devirtualization crosses
+    // servicing bubble boundaries.
+    if (derivedMethod == nullptr)
+    {
+        JITDUMP("--- no derived method, sorry\n");
+        return;
+    }
+
+    // Fetch method attributes to see if method is marked final.
+    const DWORD derivedMethodAttribs = info.compCompHnd->getMethodAttribs(derivedMethod);
+    const bool  derivedMethodIsFinal = ((derivedMethodAttribs & CORINFO_FLG_FINAL) != 0);
+
+#if defined(DEBUG)
+    const char* const derivedMethodNote = derivedMethodIsFinal ? " [final]" : "";
+    const char*       derivedClassName;
+    const char* const derivedMethodName = eeGetMethodName(derivedMethod, &derivedClassName);
+    if (verbose)
+    {
+        printf("    devirt to %s::%s%s\n", derivedClassName, derivedMethodName, derivedMethodNote);
+        gtDispTree(call);
+    }
+#endif // defined(DEBUG)
+
+    if (!isExact && !objClassIsFinal && !derivedMethodIsFinal)
+    {
+        // Type is not exact, and neither class or method is final.
+        //
+        // We could speculatively devirtualize, but there's no
+        // reason to believe the derived method is the one that
+        // is likely to be invoked.
+        //
+        // If there's currently no further overriding (that is, at
+        // the time of jitting, objClass has no subclasses that
+        // override this method), then perhaps we'd be willing to
+        // make a bet...?
+        JITDUMP("    Class NOT final or exact, no devirtualization\n");
+        return;
+    }
+
+    JITDUMP("!!! %s; can devirtualize\n", isExact ? "exact type known" : "final class or method");
+
+    // Make the updates.
+    call->gtFlags &= ~GTF_CALL_VIRT_VTABLE;
+    call->gtFlags &= ~GTF_CALL_VIRT_STUB;
+    call->gtCallMethHnd = derivedMethod;
+    call->gtCallType    = CT_USER_FUNC;
+
+    // Virtual calls include an implicit null check, which we may
+    // now need to make explicit. Not sure yet if we can restrict
+    // this to just a subset or need to do it for all cases, so
+    // will do it for all unless we know the object is not null.
```

Comment:
Supposing we could restrict this to a subset, would that be up in the code that's setting `nullCheck`, or down here?  If it would be down here, maybe better to invert the sense of the bool and name it `knownNotNull` or something rather than `nullCheck`.